zerver/lib/message.py

```diff
@@ -1697,3 +1698,136 @@ def update_to_dict_cache(
+    """
+    if (
+        sender.automatically_follow_topics_policy
+        == UserProfile.AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
+    if (
+        is_stream_muted
+        and sender.automatically_unmute_topics_in_muted_streams_policy
+        == UserProfile.AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.UNMUTED
+
+    return None
+
+
+def visibility_policy_for_send_message(
+    sender: UserProfile,
+    message: Message,
+    is_stream_muted: Optional[bool],
+    current_visibility_policy: int,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a message
+    is sent to a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_in_muted_streams_policy' settings.
+
+    It returns None when the policies can't make it more visible than the
+    current visibility policy.
+    """
+    # We prioritize 'FOLLOW' over 'UNMUTE' in muted streams.
+    # We need to carefully handle the following two cases:
+    #
+    # 1. When a user starts a topic
+    #    note: starting a topic is INITIATION as well as PARTICIPATION
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_in_muted_streams_policy' | visibility_policy
+    #       ON_PARTICIPATION               |                   ON_INITIATION                       |     FOLLOWED
+    #         ON_INITIATION                |                 ON_PARTICIPATION                      |     FOLLOWED
+    #         ON_INITIATION                |                   ON_INITIATION                       |     FOLLOWED
+    #
+    # 2. When both the policies have the same value 'ON_PARTICIPATION'.
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_in_muted_streams_policy' | visibility_policy
+    #       ON_PARTICIPATION               |                 ON_PARTICIPATION                       |     FOLLOWED
+    visibility_policy = None
+
+    if current_visibility_policy == UserTopic.VisibilityPolicy.FOLLOWED:
+        return visibility_policy
+
+    visibility_policy = visibility_policy_for_participation(sender, is_stream_muted)
+
+    if visibility_policy == UserTopic.VisibilityPolicy.FOLLOWED:
+        return visibility_policy
+
+    # If a topic has a visibility policy set, it can't be the case
+    # of initiation. We return early, thus saving a DB query.
+    if current_visibility_policy != UserTopic.VisibilityPolicy.INHERIT:
+        if visibility_policy and current_visibility_policy == visibility_policy:
+            return None
+        return visibility_policy
+
+    old_accessible_messages_in_topic = UserMessage.objects.filter(
+        user_profile=sender,
+        message__recipient_id=message.recipient_id,
+        message__subject__iexact=message.topic_name(),
+    ).exclude(message_id=message.id)
```

Comment:
Reading this check, I realized that I think it is wrong for in the case of a stream with shared history; in that case, 
the initiation check needs to be more like `get_topic_history_for_public_stream` not `get_topic_history_for_stream` -- basically we need a `can_access_stream_history` if/else clause here, and filter on just `Message` if the user has history access.

Reply:
Updated code:

```python
    # Now we need to check if the user initiated the topic.
    old_accessible_messages_in_topic: Union[QuerySet[Message], QuerySet[UserMessage]]
    if can_access_stream_history(sender, stream):
        old_accessible_messages_in_topic = messages_for_topic(
            realm_id=sender.realm_id,
            stream_recipient_id=message.recipient_id,
            topic_name=message.topic_name(),
        ).exclude(id=message.id)
    else:
        # We use the user's own message access to avoid leaking information in
        # private streams with protected history.
        old_accessible_messages_in_topic = UserMessage.objects.filter(
            user_profile=sender,
            message__recipient_id=message.recipient_id,
            message__subject__iexact=message.topic_name(),
        ).exclude(message_id=message.id)
```